### PR TITLE
Add IRemoteJobManager interface

### PR DESCRIPTION
**Pull request by FedeMPouzols**
This addresses ticket [#11123](http://trac.mantidproject.org/mantid/ticket/11123) and it is an intermediate step in a chain of tickets towards generic remote algorithms. This just adds an interface. It will become more testable when the RemoteJobManager factory and the specialized job managers are added in follow-up PRs.

To test: this PR just adds an interface. See the ticket for details on the "remote job manager" design. Check that it is well documented and that the documentation reads well. As this goes in principle in Framework/Kernel, it would be great if an experienced Mantid developer double checks that everything is fine and follows the rules/conventions/style.

### CODE REVIEW

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

File: Code/Mantid/Framework/Kernel/inc/MantidKernel/IRemoteJobManager.h [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+   * @param username User name or credentials
+   *
+   * @param password Password (or other type of authentication token)
+   * string.
+   *
+   * @throws std::invalid_argument If any of the inputs is not set
+   * properly.
+   * @throws std::runtime_error If authentication fails
+   */
+  virtual void authenticate(std::string &username, std::string &password) = 0;
```
Comment (martyngigg) [inline Code/Mantid/Framework/Kernel/inc/MantidKernel/IRemoteJobManager.h]: Is there a good reason that the parameters are non-const references when all other methods in the interface accept a const reference to a string?

Comment (FedeMPouzols) [inline Code/Mantid/Framework/Kernel/inc/MantidKernel/IRemoteJobManager.h]: Ups, I'll fix it, there's definitely no reason.

**Comment by martyngigg** [comment]
Most of our interface classes are in API, is there a reason this is in Kernel?

**Comment by FedeMPouzols** [comment]
I guess that I just put it in the same place where the current RemoteJobManager class lives. 
But yes, it doesn't make much sense to me, and actually the RemoteJobManagerFactory that comes in a follow-up ticket is inside API. So I'll move this interface into API.

**Comment by martyngigg** [comment]
Kernel/API are a somewhat arbitrary split so I just wondered on the reason. Thanks

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Seemingly unrelated and random system test failue.. Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins could you retest this please?

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

File: Code/Mantid/Framework/API/inc/MantidAPI/IRemoteJobManager.h [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+    /// ID of the transaction where this job is included
+    std::string transactionID;
+    /// Date-time of submission. No particular format can be assumed
+    std::string submitDate;
+    /// Date-time the job actually started running.  No particular
+    /// format can be assumed
+    std::string startDate;
+    /// Date-time the job finished. No particular format can be
+    /// assumed
+    std::string completionTime;
```
Comment (mantid-roman) [inline Code/Mantid/Framework/API/inc/MantidAPI/IRemoteJobManager.h]: Shouldn't the time-related have a more appropriate type (DateAndType for example)?

Comment (FedeMPouzols) [inline Code/Mantid/Framework/API/inc/MantidAPI/IRemoteJobManager.h]: Good point, I'm going to change the type of those three to DateAndTime.
The schedulers that we have in practice (LSF and Mantid API/MOAB) use different formats but both can be converted to DateAndTime.

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins, retest this please

**Comment by FedeMPouzols** [comment]
Jenkins... retest this please

**Pull request merged**